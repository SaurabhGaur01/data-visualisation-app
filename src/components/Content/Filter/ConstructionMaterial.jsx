import React from 'react';
import { Bar } from 'react-chartjs-2';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { getConstructionMaterialMatrix, getAreaSizeMatrix } from '../../../utils/transformedUtils';
import { FILTER_BY_CONSTRUCTION_MATERIAL } from '../../../constants/filterOptions';
import {
    CHART_BACKGROUND_COLOR,
    CHART_OPTION,
    CONSTRUCTION_MATERIAL_TITLE,
    CATEGORY_TITLE,
} from '../../../constants/chart';
import handleSelectedDataSet from '../../../thunks/handleSelectedDataSet';

const ConstructionMaterial = ({
    filteredMaterial,
    filterBy,
    actionHandleSelectedDataSet,
}) => {
    var showMatrix = [];
    var labelMatrix = [];
    var valueMatrix = [];
    var title = '';
    if(filterBy === FILTER_BY_CONSTRUCTION_MATERIAL) {
        showMatrix = getConstructionMaterialMatrix(filteredMaterial);
        Object.entries(showMatrix).forEach(([key, value]) => {
            labelMatrix.push(key);
            valueMatrix.push(value);
        })
        title = CONSTRUCTION_MATERIAL_TITLE;
    } else {
        showMatrix = getAreaSizeMatrix(filteredMaterial);
        Object.entries(showMatrix).forEach(([key, value]) => {
            labelMatrix.push(key);
            valueMatrix.push(value);
        })
        title = CATEGORY_TITLE;
    }

    const data = {
      labels: labelMatrix,
      datasets: [
        {
            label: title,
            data: valueMatrix,
            backgroundColor: CHART_BACKGROUND_COLOR,
            borderWidth: 2,  
        },
      ],
    }
  
    const getElementAtEvent = element => {
        if (!element.length) return;
        const { _model: { label } } = element[0];
        actionHandleSelectedDataSet(label);
    };

    return (
        <div className="chart-container">
          <Bar 
            data={data} 
            options={CHART_OPTION} 
            getElementAtEvent={getElementAtEvent}
          />
        </div>
    );
}

ConstructionMaterial.propTypes = {
    filterBy: PropTypes.string.isRequired,
    filteredMaterial: PropTypes.arrayOf(PropTypes.shape({})),
    actionHandleSelectedDataSet: PropTypes.func.isRequired,
};
  
export const mapStateToProps = state => ({
    filterBy: state.filterBy,
    filteredMaterial: state.filteredMaterial,
});

const mapDispatchAsProps = {
    actionHandleSelectedDataSet: handleSelectedDataSet,
}

const hocChain = compose(
    connect(mapStateToProps, mapDispatchAsProps),
);

export { ConstructionMaterial as TestableConstructionMaterial };

export default hocChain(ConstructionMaterial);