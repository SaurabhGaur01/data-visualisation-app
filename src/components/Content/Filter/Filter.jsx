import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { compose } from "redux";
import { setFilterBy } from "../../../ducks/filterBy";
import {
  FILTER_BY_CONSTRUCTION_MATERIAL,
  FILTER_BY_SIZE_CATEGORY,
  FILTER_BY_NONE,
} from "../../../constants/filterOptions";
import handleReset from "../../../thunks/handleReset";

const Filter = ({ actionSetFilterBy, actionHandleReset, filterBy }) => {
  const changeFilter = (event) => {
    actionSetFilterBy(event.target.value);
    if (event.target.value === FILTER_BY_NONE) {
      actionHandleReset();
    }
  };

  return (
    <div className="filter-by mr-5">
      <span className="filter-heading pr-2">Filter By</span>
      <select
        id="filter-by"
        name="filter-by"
        className="input"
        onChange={changeFilter}
        value={filterBy}
      >
        <option value={FILTER_BY_CONSTRUCTION_MATERIAL}>Material</option>
        <option value={FILTER_BY_SIZE_CATEGORY}>Size</option>
        <option value={FILTER_BY_NONE}>None</option>
      </select>
    </div>
  );
};

Filter.propTypes = {
  actionSetFilterBy: PropTypes.func.isRequired,
  actionHandleReset: PropTypes.func.isRequired,
  filterBy: PropTypes.string.isRequired,
};

export const mapStateToProps = (state) => ({
  filterBy: state.filterBy,
});

const mapDispatchAsProps = {
  actionSetFilterBy: setFilterBy,
  actionHandleReset: handleReset,
};

const hocChain = compose(
  connect(mapStateToProps, mapDispatchAsProps)
);

export { Filter as TestableFilter };

export default hocChain(Filter);
