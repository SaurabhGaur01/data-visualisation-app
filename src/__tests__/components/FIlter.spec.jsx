import React from 'react';
import { shallow } from 'enzyme';
import { TestableFilter, mapStateToProps} from '../../components/Content/Filter/Filter';

describe('Filter Component', () => {
    let props;

    beforeEach(() =>{
        props = {
            actionHandleReset: jest.fn(),
            actionSetFilterBy: jest.fn(),
            filterBy: 'mockFilterBy',
        };
    });

    it('should render correctly', () => {
        const renderedModule = shallow(<TestableFilter {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    it('should call actionSetFilterBy when dropdown value changed but not none', () => {
        const renderedModule = shallow(<TestableFilter {...props} />);
        renderedModule.find('#filter-by').at(0).simulate('change', {target: { value: 'mockValue'}});
        expect(props.actionSetFilterBy).toHaveBeenCalledTimes(1);
        expect(props.actionSetFilterBy).toHaveBeenCalledWith('mockValue');
    });

    it('should call actionHandleReset when dropdown value changed to none', () => {
        const renderedModule = shallow(<TestableFilter {...props} />);
        renderedModule.find('#filter-by').at(0).simulate('change', {target: { value: 'none'}});
        expect(props.actionHandleReset).toHaveBeenCalledTimes(1);
    });

    describe('mapStateToProps', () => {        
        const mockedState = {
            filterBy: 'mockFilterBy',
        };

        it('should return the module state from the redux store', () => {
            const mockPropsFromRedux = mapStateToProps(mockedState);
            expect(mockPropsFromRedux).toEqual(
                {
                    filterBy: 'mockFilterBy',
                }
            )
        })
    })
});    