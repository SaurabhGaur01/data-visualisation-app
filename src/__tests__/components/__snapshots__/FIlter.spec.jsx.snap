// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Component should render correctly 1`] = `
<div
  className="filter-by mr-5"
>
  <span
    className="filter-heading pr-2"
  >
    Filter By
  </span>
  <select
    className="input"
    id="filter-by"
    name="filter-by"
    onChange={[Function]}
    value="mockFilterBy"
  >
    <option
      value="construction"
    >
      Material
    </option>
    <option
      value="category"
    >
      Size
    </option>
    <option
      value="none"
    >
      None
    </option>
  </select>
</div>
`;
